tools/trace2syz/trace2syz/intermediateTypes.go

```diff
@@ -0,0 +1,686 @@
+}
+
+func newCallType(name string, args []irType) (typ *call) {
+	typ = new(call)
+	typ.CallName = name
+	typ.Args = args
+	return
+}
+
+func (c *call) Name() string {
+	return "Call Type"
+}
+
+func (c *call) String() string {
+	var buf bytes.Buffer
+
+	buf.WriteString("Name: " + c.CallName + "\n")
+	for _, arg := range c.Args {
+		buf.WriteString("Arg: " + arg.Name() + "\n")
+	}
+	return buf.String()
+}
+
+func (c *call) Eval(target *prog.Target) uint64 {
+	panic("Eval called on call type")
+}
+
+type binop struct {
+	Operand1 *expression
+	Op       Operation
+	Operand2 *expression
+}
+
+func newBinop(operand1 irType, op Operation, operand2 irType) (b *binop) {
+	b = new(binop)
+	b.Operand1 = operand1.(*expression)
+	b.Op = op
+	b.Operand2 = operand2.(*expression)
+	return
+}
+
+func (b *binop) Eval(target *prog.Target) uint64 {
+	op1Eval := b.Operand1.Eval(target)
+	op2Eval := b.Operand2.Eval(target)
+	switch b.Op {
+	case ANDop:
+		return op1Eval & op2Eval
+	case ORop:
+		return op1Eval | op2Eval
+	case XORop:
+		return op1Eval ^ op2Eval
+	case LSHIFTop:
+		return op1Eval << op2Eval
+	case RSHIFTop:
+		return op1Eval >> op2Eval
+	case TIMESop:
+		return op1Eval * op2Eval
+	default:
+		panic("Operator Not handled")
+	}
+}
+
+func (b *binop) String() string {
+	return fmt.Sprintf("op1: %s op2: %s, operand: %v\n", b.Operand1.String(), b.Operand2.String(), b.Op)
+}
+
+func (b *binop) Name() string {
+	return "Binop"
+}
+
+type unop struct {
+	Operand *expression
+	Op      Operation
+}
+
+func newUnop(operand irType, op Operation) (u *unop) {
+	u = new(unop)
+	u.Operand = operand.(*expression)
+	u.Op = op
+	return
+}
+
+func (u *unop) Eval(target *prog.Target) uint64 {
+	opEval := u.Operand.Eval(target)
+	switch u.Op {
+	case ONESCOMPop:
+		return ^opEval
+	default:
+		panic("Unsupported Unop Op")
+	}
+}
+
+func (u *unop) String() string {
+	return fmt.Sprintf("op1: %v operand: %v\n", u.Operand, u.Op)
+}
+
+func (u *unop) Name() string {
+	return "Unop"
+}
+
+type field struct {
+	Key string
+	Val irType
+}
+
+func newField(key string, val irType) (f *field) {
+	f = new(field)
+	f.Key = key
+	f.Val = val
+	return
+}
+
+func (f *field) Name() string {
+	return "Field Type"
+}
+
+func (f *field) String() string {
+	return f.Val.String()
+}
+
+func (f *field) Eval(target *prog.Target) uint64 {
+	return f.Val.Eval(target)
+}
+
+type intType struct {
+	Val int64
+}
+
+func newIntsType(vals []int64) ints {
+	ints := make([]*intType, 0)
+	for _, v := range vals {
+		ints = append(ints, newIntType(v))
+	}
+	return ints
+}
+
+func newIntType(val int64) (typ *intType) {
+	typ = new(intType)
+	typ.Val = val
+	return
+}
+
+func (i *intType) Eval(target *prog.Target) uint64 {
+	return uint64(i.Val)
+}
+
+func (i *intType) Name() string {
+	return "Int Type"
+}
+
+func (i *intType) String() string {
+	return strconv.FormatInt(i.Val, 10)
+}
+
+type flags []*flagType
+
+type ints []*intType
+
+func (f flags) Eval(target *prog.Target) uint64 {
+	if len(f) > 1 {
```

Comment:
It seems that all of the >1, ==1, else cases are unnecessary here. The loop for >1 will handle all of these cases.

Reply:
ack